api/turing/cluster/controller_test.go

```diff
@@ -1737,8 +1738,19 @@ func TestApplyIstioVirtualService(t *testing.T) {
 					resource: virtualServiceResource.Resource,
 					rFunc: func(action k8stesting.Action) (bool, runtime.Object, error) {
 						expAction := k8stesting.NewCreateAction(virtualServiceResource, testNamespace, testVs)
+						// expAction.Object.Spec.state.atomicMessageInfo is nil, unlike action.Object.Spec.state.atomicMessageInfo
+						// we thus need to check each of the other field manually
 						// Check that the method is called with the expected action
-						assert.Equal(t, expAction, action)
+						actualAction, ok := action.(k8stesting.CreateAction)
+						assert.True(t, ok)
+						actual, ok := actualAction.GetObject().(*v1beta1.VirtualService)
+						assert.True(t, ok)
+						expected, ok := expAction.GetObject().(*v1beta1.VirtualService)
+						assert.True(t, ok)
+
+						assert.Equal(t, expected.ObjectMeta, actual.ObjectMeta)
+						assert.Equal(t, expected.Spec.String(), actual.Spec.String())
+						assert.Equal(t, expected.Status.String(), actual.Status.String())
```

Comment:
Similar to the comment [here](https://github.com/caraml-dev/turing/pull/371/files#r1555448202), we cannot compare expected and got directly since VirtualService-s returned by the Istio client now contain unexported fields as well as mutex locks. In order to work around this, we can compare the string generated from the VirtualService's spec and status that prints out most of the fields in these fields.